h2o-algos/src/main/java/hex/tree/SharedTreeModelGBMDRF.java

```diff
@@ -0,0 +1,159 @@
+import water.*;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+import water.fvec.NewChunk;
+import water.fvec.Vec;
+import water.util.*;
+
+import java.util.ArrayList;
+import java.util.List;
+import hex.Model;
+import water.Key;
+
+public abstract class SharedTreeModelGBMDRF<
+        M extends SharedTreeModel<M, P, O>,
+        P extends SharedTreeModel.SharedTreeParameters,
+        O extends SharedTreeModel.SharedTreeOutput
+        > extends SharedTreeModel<M, P, O> implements Model.Contributions{
+    
+    public SharedTreeModelGBMDRF(Key<M> selfKey, P parms, O output) {
+        super(selfKey, parms, output);
+    }
+    
+  @Override
+  public Frame scoreContributions(Frame frame, Key<Frame> destination_key) {
+    if (_output.nclasses() > 2) {
+      throw new UnsupportedOperationException(
+              "Calculating contributions is currently not supported for multinomial models.");
+    }
+
+    Frame adaptFrm = new Frame(frame);
+    adaptTestForTrain(adaptFrm, true, false);
+    // remove non-feature columns
+    adaptFrm.remove(_parms._response_column);
+    adaptFrm.remove(_parms._fold_column);
+    adaptFrm.remove(_parms._weights_column);
+    adaptFrm.remove(_parms._offset_column);
+
+    final String[] outputNames = ArrayUtils.append(adaptFrm.names(), "BiasTerm");
+    
+    if (_parms.algoName().equals("DRF")) {
+      return new ScoreContributionsTaskDRF(this, _output._ntrees, _output._treeKeys, _output._init_f)
+              .doAll(outputNames.length, Vec.T_NUM, adaptFrm)
+              .outputFrame(destination_key, outputNames, null);
+    } else {
+      return new ScoreContributionsTask(this, _output._ntrees, _output._treeKeys, _output._init_f)
+              .doAll(outputNames.length, Vec.T_NUM, adaptFrm)
+              .outputFrame(destination_key, outputNames, null);
+    }
+  }
+  
+  public class ScoreContributionsTaskDRF extends ScoreContributionsTask {
+
+    public ScoreContributionsTaskDRF(SharedTreeModel model, int ntrees, Key<CompressedTree>[][] treeKeys, double init_f) {
+      super(model, ntrees, treeKeys, init_f);
+    }
+    
+    @Override
+    public void map(Chunk chks[], NewChunk[] nc) {
+      assert chks.length == nc.length - 1; // calculate contribution for each feature + the model bias
+      double[] input = MemoryManager.malloc8d(chks.length);
+      float[] contribs = MemoryManager.malloc4f(nc.length);
+
+      Object workspace = _treeSHAP.makeWorkspace();
+
+      for (int row = 0; row < chks[0]._len; row++) {
+        for (int i = 0; i < chks.length; i++) {
+          input[i] = chks[i].atd(row);
+        }
+        for (int i = 0; i < contribs.length; i++) {
+          contribs[i] = 0;
+        }
+
+        // calculate Shapley values
+        _treeSHAP.calculateContributions(input, contribs, 0, -1, workspace);
+        
+        for (int i = 0; i < nc.length; i++) {
+          // Prediction of DRF tree ensemble is an average prediction of all trees. So, divide contribs by ntrees
```

Comment:
only the inside of the for loop is different, it should be an abstract method in ScoreContributionsTask with implementation in DRF and GBM model classes

Reply:
Yes, just pushed a change for this: https://github.com/h2oai/h2o-3/pull/3628/commits/ddb6a071d58f62d0f64eee8b6174d34874eacc1f